### PR TITLE
added 4x12

### PR DESCRIPTION
**Pull request by daysgobye**
## Description
added keyboard 4x12 to the boardsource folder

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/boardsource/4x12/rules.mk [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+MOUSEKEY_ENABLE = yes       # Mouse keys(+4700)
+EXTRAKEY_ENABLE = yes       # Audio control and System control(+450)
+CONSOLE_ENABLE = yes        # Console for debug(+400)
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+MIDI_ENABLE = no            # MIDI support (+2400 to 4200, depending on config)
```
Comment (fauxpark) [inline keyboards/boardsource/4x12/rules.mk:27]:
```suggestion
BOOTMAGIC_ENABLE = lite      # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = yes        # Console for debug
COMMAND_ENABLE = yes        # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
MIDI_ENABLE = no            # MIDI support
```

File: keyboards/boardsource/4x12/4x12.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+#include "quantum.h"
+
+#define LAYOUT( \
```
Comment (drashna) [inline keyboards/boardsource/4x12/4x12.h:4]:
```suggestion
#define LAYOUT_ortho_4x12( \
```

File: keyboards/boardsource/4x12/info.json [hunk 3]
```diff
@@ -0,0 +1,64 @@
+{
+    "keyboard_name": "Boardsource 4x12",
+    "url": "https://boardsource.xyz",
+    "maintainer": "Boardsource",
+    "width": 12,
+    "height": 4,
+    "layouts": {
+        "LAYOUT": {
```
Comment (drashna) [inline keyboards/boardsource/4x12/info.json:8]:
```suggestion
        "LAYOUT_ortho_4x12": {
```

File: keyboards/boardsource/4x12/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    _LOWER,
+};
+
+// Readability keycodes
+#define LOWER   MO(_LOWER)
+#define RAISE   MO(_RAISE)
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+  [_MAIN] = LAYOUT(
```
Comment (drashna) [inline keyboards/boardsource/4x12/keymaps/default/keymap.c:15]:
```suggestion
  [_MAIN] = LAYOUT_ortho_4x12(
```

File: keyboards/boardsource/4x12/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_BSPC,
+    KC_ESC,  KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,
+    KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_ENT ,
+    KC_PIPE, KC_LCTL, KC_LALT, KC_LGUI, LOWER,   KC_SPC,  KC_SPC,  RAISE,   KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT
+  ),
+
+  [_RAISE] = LAYOUT(
```
Comment (drashna) [inline keyboards/boardsource/4x12/keymaps/default/keymap.c:22]:
```suggestion
  [_RAISE] = LAYOUT_ortho_4x12(
```

File: keyboards/boardsource/4x12/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  KC_TILD, KC_EXLM, KC_AT,   KC_HASH, KC_DLR,  KC_PERC, KC_CIRC, KC_AMPR,    KC_ASTR,    KC_LPRN, KC_RPRN, KC_BSPC,
+  KC_DEL,  KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_UNDS,    KC_PLUS,    KC_LCBR, KC_RCBR, KC_PIPE,
+  _______, KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,  S(KC_NUHS), S(KC_NUBS), KC_HOME, KC_END,  _______,
+  RESET,   _______, _______, _______, _______, _______, _______, _______,    KC_MNXT,    KC_VOLD, KC_VOLU, KC_MPLY
+  ),
+
+  [_LOWER] = LAYOUT(
```
Comment (drashna) [inline keyboards/boardsource/4x12/keymaps/default/keymap.c:29]:
```suggestion
  [_LOWER] = LAYOUT_ortho_4x12(
```

File: keyboards/boardsource/4x12/rules.mk [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+MIDI_ENABLE = no            # MIDI support
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
```
Comment (drashna) [inline keyboards/boardsource/4x12/rules.mk:31]:
```suggestion
FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches

LAYOUTS = ortho_4x12
```

**Comment by daysgobye** [comment]
this keyboard does not have back-lighting I and it is not enabled in the rules.mk I am not sure why the build is failing, it does not seem to fail on my computer. Should I define a unused pin as a backlight pin?

**Pull request merged**